Lib/logging/handlers.py

```diff
@@ -401,8 +402,12 @@ def doRollover(self):
                 timeTuple = time.localtime(t + addend)
         dfn = self.rotation_filename(self.baseFilename + "." +
                                      time.strftime(self.suffix, timeTuple))
-        if os.path.exists(dfn):
-            os.remove(dfn)
+        while os.path.exists(dfn):
```

Comment:
I think this would be more future-proof if it were structured as:
```python
if os.path.exists(dfn):
    self.backup_existing(dfn)
```
and the new `backup_existing` method would have the behaviour similar to this docstring:
```python
def backup_existing(self, dfn):
    """
    Back up an existing file with the same name `dfn` as a destination file for rollover.

    This method must find a unique filename in the same directory, related to `dfn`,
    and rename the existing file to that name.

    This implementation appends ' (N)' to `dfn`, where N is an integer going up from 1
    until an integer value N is found such that no existing file with that ' (N)' suffix is found.
    """
```
Having a separate method allows someone with a custom scheme for naming logfiles (using the `namer` attribute) to adapt their backup to that scheme by overriding the method.

Reply:
Sorry for being slow, I just wanted to clear make sure I have the correct understanding. I agree that the having a separate `backupExisting` method is a good idea (probably `backupExisting` instead of `backup_existing` for consistency with the rest of the module?)

I wasn't 100% sure on the behaviour of the method though, from my understanding on the rollover, the new file should be created with a suffix, from the docstring above, it seems like you want to rename the existing file to have a suffix?

as an example, I have the following log files
```
test.log
========
[INFO]: foo

test.log.2021-08-02
========
[INFO]: bar
```

for all the examples below, I'll denote that structure as
```
test.log # foo
test.log.2021-08-02 # bar
```

I was imagining the behaviour below:
```
test.log # baz
test.log.2021-08-02 # bar
test.log.2021-08-02 (1) # foo
```

This would mean that for n files, the latest file would always have the highest suffix. The current log would obviously have no suffix, the oldest rollover log would only have the date, and the following files would have (1), ... (N) in the order that they were generated.

From reading your comment, implemeting that method would yield the behaviour below:
```
test.log # baz
test.log.2021-08-02 # foo
test.log.2021-08-02 (1) # bar
```
where the most recent file would rename the already existing file in the directory and "bump" it up a number. 

Personally I prefer the first method as implementing the method this way could cause some ambiguity i.e. in the case of a directory with suffixes `(1)`, `(2)`, `(3)`, what happens to the newest file? The newest file gets the suffix `(1)`, and then does the already existing `(1)` get bumped to `(4)`, or does it get bumped to `(2)`, and then `(2)` -> `(3)` and so on.

I'm not sure how this different logic affects the other issues you linked, I had a look at them but I need to do some testing to see the results myself as I couldn't 100% follow just the PR and bpo. I'll try and implement the logic and update this PR tonight.

Reply:
Don't rush to update the PR, because maybe we should think about things some more. I'm not able to say much at the minute as I'm pressed for time, but I was thinking along the lines of: a target file that exists when rolling over should be an unusual case, not a common case (This issue hasn't come up in the many years since `TimedRotatingFileHandler` was added, IIRC). So, my proposed approach treats the existing file as just a backup (so stuff doesn't get lost), but once it's renamed, it's cleared the way for the expected `dfn` to be used, and the "usual" scheme applies from then on for log files.

Reply:
> (probably backupExisting instead of backup_existing for consistency with the rest of the module?

Yes, good point. The original logging module was written back before PEP8, and kept the old-style naming for backward compatibility. So `backupExisting` is right.